src/lightning/data/cache/writer.py

```diff
@@ -0,0 +1,273 @@
+            "samples": len(self._serialized_items),
+            "filename": filename,
+            "mapping": mapping,
+            "interval": [self._indexes[0], self._indexes[-1] + 1],
+        }
+
+        self._chunks_info.append(chunk_info)
+
+        return data
+
+    def write_chunk(self):
+        """Write a chunk to the filesystem."""
+        if self._compression:
+            filename = f"chunk-{self.rank}-{self._chunk_id}.{self._compression}.bin"
+        else:
+            filename = f"chunk-{self.rank}-{self._chunk_id}.bin"
+        self.write_file(self._create_chunk(filename), filename)
+
+    @property
+    def available_serializers(self):
+        return self._serializers
+
+    def reset(self) -> None:
+        """Reset the writer to handle the next chunk."""
+        self._serialized_items = []
+        self._indexes = []
+        self._current_chunk_size = 0
+
+    def __setitem__(self, index, items: any):
+        """Store an item to a chunk.
+
+        The index needs to be provided in order.
+
+        """
+        serialized_items = self.serialize(items)
+        serialized_items_size = len(serialized_items)
+
+        if self._chunk_size < self._current_chunk_size + serialized_items_size:
+            if self._current_chunk_size == 0:
+                raise Exception(
+                    f"The provided chunk_size {self._chunk_size} is too small."
+                    f" You should use a multiple of {serialized_items_size} bytes."
+                )
+            self.write_chunk()
+            self.reset()
+            self._chunk_id += 1
+
+        self._serialized_items.append(serialized_items)
+        self._current_chunk_size += serialized_items_size
+
+        if self._indexes:
+            assert self._indexes[-1] == index - 1
+
+        self._indexes.append(index)
+
+    def write_file(
+        self,
+        raw_data: bytes,
+        filename: str,
+    ) -> None:
+        """Write chunk bytes to a file."""
+        if self._compression:
+            raw_data = self._compressor.compress(raw_data)
+        filepath = os.path.join(self._cache_dir, filename)
+        with open(filepath, "wb") as out:
+            out.write(raw_data)
+
+    def write_chunks_index(self):
+        """Write the chunks index to a JSON file."""
+        filepath = os.path.join(self._cache_dir, f"{self.rank}.index.json")
+        config = self.get_config()
+        with open(filepath, "w") as out:
+            json.dump({"chunks": self._chunks_info, "config": config}, out, sort_keys=True)
+
+    def done(self):
+        """Called when StopIteration is triggered.
+
+        It tries to save the last chunk and write the chunks index.
+
+        """
+        if self.filled:
+            return
+        if self._serialized_items:
+            self.write_chunk()
+        self.write_chunks_index()
+        self.reset()
+        self._is_done = True
+
+    @classmethod
+    def get_cloud_path(cls, cache_dir: str) -> Optional[str]:
```

Comment:
Should we keep this stuff here? Maybe we can plug it from outside?

Reply:
Sure. This is mainly for automation.

Reply:
Moved to its own function.